File: src/main/java/me/dkim19375/dkim19375core/external/PAPIExpansion.java
package me.dkim19375.dkim19375core.external;

import me.clip.placeholderapi.expansion.PlaceholderExpansion;
import org.bukkit.OfflinePlayer;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class PAPIExpansion extends PlaceholderExpansion {
    private final String identifier;
    private final String authors;
    private final String version;

    public PAPIExpansion(final @NotNull JavaPlugin plugin, final @Nullable String identifier, final @Nullable String authors, final @Nullable String version) {
        if (identifier == null) {
            this.identifier = plugin.getDescription().getName();
        } else {
            this.identifier = identifier;
        }
        if (authors == null) {
            this.authors = plugin.getDescription().getAuthors().toString();
        } else {
            this.authors = authors;
        }
        if (version == null) {
            this.version = plugin.getDescription().getVersion();
        } else {
            this.version = version;
        }
    }

    @Override
    public boolean persist() {
        return true;
    }

    @Override
    public boolean canRegister() {
        return true;
    }

    @Override
    @NotNull
    public String getAuthor() {
        return authors;
    }

    @Override
    @NotNull
    public String getIdentifier() {
        return identifier;
    }

    @Override
    @NotNull
    public String getVersion() {
        return version;
    }

    @Override
    @Nullable
    public String onPlaceholderRequest(final Player player, final @Nullable String identifier) {
        return null;
    }

    @Override
    @Nullable
    public String onRequest(OfflinePlayer player, @NotNull String params) {
        return null;
    }
}

File: src/main/kotlin/me/dkim19375/dkim19375core/external/PAPIExpansion.kt
package me.dkim19375.dkim19375core.external

import me.clip.placeholderapi.expansion.PlaceholderExpansion
import me.dkim19375.dkim19375core.annotation.API
import me.dkim19375.dkim19375core.javaplugin.CoreJavaPlugin

@API
abstract class PAPIExpansion(
    private val plugin: CoreJavaPlugin,
    private val identifier: String = plugin.name,
    private val authors: String = plugin.description.authors.joinToString(", "),
    private val version: String = plugin.description.version
) : PlaceholderExpansion() {

    override fun persist(): Boolean = true

    override fun canRegister(): Boolean = true

    override fun getAuthor(): String = authors

    override fun getIdentifier(): String = identifier

    override fun getVersion(): String = version
}